File: khandaq-android-trifa/android-refimpl-app/app/src/main/java/com/zoffcc/applications/trifa/ReconnectBackoffCoordinator.java
package com.zoffcc.applications.trifa;

import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;

import java.util.Random;
import java.util.concurrent.atomic.AtomicInteger;

import static com.zoffcc.applications.trifa.NetworkDiagnosticsLog.log;
import static com.zoffcc.applications.trifa.TrifaToxService.on_network_reconnect_attempt;

/**
 * Exponential backoff with jitter for DHT rebootstrap (1s, 2s, 4s … max 30s).
 * Urgent network-change events fire immediately (debounce handled upstream).
 */
final class ReconnectBackoffCoordinator
{
    private static final String TAG = "trifa.ReconnectBackoff";
    private static final long BASE_DELAY_MS = 1000L;
    private static final long MAX_DELAY_MS = 30_000L;
    private static final int MAX_ATTEMPTS = 12;

    enum Reason
    {
        NETWORK_CHANGE,
        OFFLINE_PERIODIC,
        BACKGROUND_WAKE,
        MANUAL
    }

    private static ReconnectBackoffCoordinator instance;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private final Random jitter = new Random();
    private final AtomicInteger attempt = new AtomicInteger(0);
    private Runnable pending;
    private long lastSuccessMs = 0L;

    static ReconnectBackoffCoordinator get()
    {
        if (instance == null)
        {
            instance = new ReconnectBackoffCoordinator();
        }
        return instance;
    }

    void onConnectionRestored()
    {
        handler.post(() -> {
            if (attempt.get() > 0)
            {
                log("reconnect_ok", "attempts_reset=" + attempt.get());
            }
            attempt.set(0);
            lastSuccessMs = SystemClock.elapsedRealtime();
            cancelPending();
        });
    }

    void scheduleReconnect(final Reason reason, final boolean urgent)
    {
        handler.post(() -> scheduleReconnectOnHandler(reason, urgent));
    }

    private void scheduleReconnectOnHandler(final Reason reason, final boolean urgent)
    {
        cancelPending();

        if (urgent)
        {
            attempt.set(0);
        }

        final int currentAttempt = attempt.get();
        final long delayMs = urgent ? 0L : computeDelayMs(currentAttempt);

        log("reconnect_scheduled",
                "reason=" + reason + " urgent=" + urgent + " attempt=" + currentAttempt + " delayMs=" + delayMs);

        pending = () -> executeReconnect(reason, currentAttempt);
        if (delayMs <= 0L)
        {
            handler.post(pending);
        }
        else
        {
            handler.postDelayed(pending, delayMs);
        }
    }

    private void executeReconnect(final Reason reason, final int currentAttempt)
    {
        pending = null;

        log("reconnect_attempt", "reason=" + reason + " attempt=" + currentAttempt);
        try
        {
            on_network_reconnect_attempt(reason, currentAttempt);
        }
        catch (Exception e)
        {
            log("reconnect_error", e.getMessage());
        }
    }

    void scheduleRetryIfStillOffline(final Reason reason)
    {
        handler.post(() -> {
            final int next = attempt.incrementAndGet();
            if (next >= MAX_ATTEMPTS)
            {
                log("reconnect_give_up", "max_attempts=" + MAX_ATTEMPTS);
                return;
            }
            final long delayMs = computeDelayMs(next);
            log("reconnect_retry_scheduled", "nextAttempt=" + next + " delayMs=" + delayMs);
            scheduleReconnectOnHandler(reason, false);
        });
    }

    private void cancelPending()
    {
        if (pending != null)
        {
            handler.removeCallbacks(pending);
            pending = null;
        }
    }

    static long computeDelayMs(final int attemptIndex)
    {
        final long exp = BASE_DELAY_MS * (1L << Math.min(attemptIndex, 5));
        final long capped = Math.min(exp, MAX_DELAY_MS);
        final long jitterMs = (long) (capped * 0.1 * (new Random().nextDouble()));
        return capped + jitterMs;
    }

    int getAttempt()
    {
        return attempt.get();
    }

    long getLastSuccessMs()
    {
        return lastSuccessMs;
    }
}

File: khandaq-ios/Antidote/ReconnectBackoffCoordinator.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import Foundation

/// Exponential backoff for DHT rebootstrap (Android ReconnectBackoffCoordinator parity).
final class ReconnectBackoffCoordinator {
    static let shared = ReconnectBackoffCoordinator()

    enum Reason: String {
        case networkChange = "NETWORK_CHANGE"
        case offlinePeriodic = "OFFLINE_PERIODIC"
        case manual = "MANUAL"
    }

    private static let baseDelay: TimeInterval = 1
    private static let maxDelay: TimeInterval = 30
    private static let maxAttempts = 12

    private var attempt = 0
    private var pendingWork: DispatchWorkItem?

    private init() {}

    func onConnectionRestored() {
        pendingWork?.cancel()
        pendingWork = nil
        if attempt > 0 {
            NetworkDiagnosticsLog.log("reconnect_ok", detail: "attempts_reset=\(attempt)")
        }
        attempt = 0
    }

    func scheduleReconnect(reason: Reason, urgent: Bool) {
        pendingWork?.cancel()

        if urgent {
            attempt = 0
        }

        let currentAttempt = attempt
        let delay = urgent ? 0 : Self.computeDelay(attemptIndex: currentAttempt)

        NetworkDiagnosticsLog.log("reconnect_scheduled",
                                  detail: "reason=\(reason.rawValue) urgent=\(urgent) attempt=\(currentAttempt) delay=\(delay)")

        let work = DispatchWorkItem { [weak self] in
            self?.executeReconnect(reason: reason, attempt: currentAttempt)
        }
        pendingWork = work

        if delay <= 0 {
            DispatchQueue.main.async(execute: work)
        } else {
            DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: work)
        }
    }

    private func executeReconnect(reason: Reason, attempt currentAttempt: Int) {
        pendingWork = nil
        NetworkDiagnosticsLog.log("reconnect_attempt", detail: "\(reason.rawValue) attempt=\(currentAttempt)")

        NotificationCenter.default.post(name: .khandaqManualReconnectRequested, object: reason)

        DispatchQueue.main.asyncAfter(deadline: .now() + 4) { [weak self] in
            guard let self = self else { return }
            guard let toxManager = ActiveSessionCoordinator.sharedToxManagerForReconnect() else { return }

            if toxManager.user.connectionStatus == .none {
                NetworkDiagnosticsLog.log("reconnect_still_offline", detail: reason.rawValue)
                self.scheduleRetryIfStillOffline(reason: reason)
            }
        }
    }

    func scheduleRetryIfStillOffline(reason: Reason) {
        attempt += 1
        if attempt >= Self.maxAttempts {
            NetworkDiagnosticsLog.log("reconnect_give_up", detail: "max=\(Self.maxAttempts)")
            return
        }

        let delay = Self.computeDelay(attemptIndex: attempt)
        NetworkDiagnosticsLog.log("reconnect_retry_scheduled", detail: "next=\(attempt) delay=\(delay)")

        let work = DispatchWorkItem { [weak self] in
            self?.scheduleReconnect(reason: reason, urgent: false)
        }
        pendingWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: work)
    }

    private static func computeDelay(attemptIndex: Int) -> TimeInterval {
        let exp = baseDelay * pow(2, Double(min(attemptIndex, 5)))
        let capped = min(exp, maxDelay)
        return capped + Double.random(in: 0...(capped * 0.1))
    }
}

extension Notification.Name {
    static let khandaqManualReconnectRequested = Notification.Name("KhandaqManualReconnectRequested")
}

extension ActiveSessionCoordinator {
    private static weak var reconnectCoordinator: ActiveSessionCoordinator?

    static func sharedToxManagerForReconnect() -> OCTManager? {
        return reconnectCoordinator?.toxManager
    }

    func registerForReconnectBackoff() {
        ActiveSessionCoordinator.reconnectCoordinator = self

        NotificationCenter.default.addObserver(self,
                                               selector: #selector(handleManualReconnect(_:)),
                                               name: .khandaqManualReconnectRequested,
                                               object: nil)
    }

    @objc private func handleManualReconnect(_ notification: Notification) {
        toxManager.bootstrap.rebootstrapOnNetworkChange()
    }
}
